x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java

```diff
@@ -0,0 +1,283 @@
+import org.elasticsearch.xpack.sql.qa.jdbc.JdbcIntegrationTestCase;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.sql.ResultSet;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+import static java.lang.String.join;
+import static java.util.Arrays.asList;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static org.elasticsearch.common.collect.Tuple.tuple;
+import static org.hamcrest.collection.IsEmptyCollection.empty;
+
+/**
+ * <p>This test was introduced because of the inconsistencies regarding NULL argument handling. NULL literal vs NULL field
+ * value as function arguments in some case result in different function return values.</p>
+ *
+ * <p>Functions should return with the same value no matter if the argument(s) came from a field or from a literal.</p>
+ *
+ * <p>The test class based on the example function calls (and argument specifications) generates all the
+ * permutations of the function arguments (4 options per argument: value/NULL as literal/field) and tests that the
+ * function calls with the same argument values provide the same result regardless of the source (literal, field)
+ * of the arguments.</p>
+ *
+ * <p>To ignore any of the tests, add an .ignore() method call after the Fn ctors in the FUNCTION_CALLS_TO_TEST list below, like:
+ * <code> new Fn("ASCII", "foobar").ignore()</code></p>
+ */
+public class ConsistentFunctionArgHandlingIT extends JdbcIntegrationTestCase {
+
+    private static final List<Fn> FUNCTION_CALLS_TO_TEST = asList(
+        new Fn("ASCII", "foobar"),
+        new Fn("BIT_LENGTH", "foobar"),
+        new Fn("CHAR", 66),
+        new Fn("CHAR_LENGTH", "foobar").aliases("CHARACTER_LENGTH"),
+        new Fn("CONCAT", "foo", "bar"),
+        new Fn("INSERT", "foobar", 2, 3, "replacement"),
+        new Fn("LCASE", "STRING"),
+        new Fn("LEFT", "foobar", 3),
+        new Fn("LENGTH", "foobar"),
+        new Fn("LOCATE", "ob", "foobar", 1),
+        new Fn("LTRIM", "   foobar"),
+        new Fn("OCTET_LENGTH", "foobar"),
+        new Fn("POSITION", "foobar", "ob"),
+        new Fn("REPEAT", "foobar", 10),
+        new Fn("REPLACE", "foo", "o", "bar"),
+        new Fn("RIGHT", "foobar", 3),
+        new Fn("RTRIM", "foobar   "),
+        new Fn("SPACE", 5),
+        new Fn("STARTS_WITH", "foobar", "foo"),
+        new Fn("SUBSTRING", "foobar", 1, 2),
+        new Fn("TRIM", "  foobar   "),
+        new Fn("UCASE", "foobar")
+    );
+
+    private static final List<String> NON_TESTED_FUNCTIONS;
+    static {
+        try {
+            Class<?> c = ConsistentFunctionArgHandlingIT.class;
+            NON_TESTED_FUNCTIONS = Files.readAllLines(Path.of(c.getResource(c.getSimpleName() + "-non-tested-functions.txt").toURI()));
```

Comment:
It would be great if it's possible get all the functions from FunctionRegistry:
Something like: https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java#L2199

Can also be done at a later point and not as part of this PR.

Reply:
`SqlFunctionRegistry` is not available in integration tests (can add it as a dependency, but I don't think we really need it). I get the list of the functions from `SHOW FUNCTIONS` [see here](https://github.com/palesz/elasticsearch/blob/4e20cbf0f8223c5be773c25b58e97de2b03d8fa3/x-pack/plugin/sql/qa/server/single-node/src/test/java/org/elasticsearch/xpack/sql/qa/single_node/ConsistentFunctionArgHandlingIT.java#L235), that achieves pretty much the same that `SqlFunctionRegistry` would (for more see the last paragraph [here](https://github.com/elastic/elasticsearch/pull/68379#discussion_r569648252)). The purpose of the `NON_TESTED_FUNCTIONS` list is to identify functions added after writing this test class which likely we should think about adding to the tested function list / non tested function list (one-liner). Newly added = `SHOW FUNCTIONS \ FUNCTIONS_CALLS \ NON_TESTED_FUNCTIONS` should be an empty set. The `NON_TESTED_FUNCTIONS` list should shrink later once the datetime, ... functions are also added to the tested function list (separate PR).